docs/enhancements/clusterpool-inventory.md

```diff
@@ -0,0 +1,175 @@
+
+### Handling Inventory Updates
+#### Adding An Inventory
+Adding an `Inventory` to a ClusterPool which previously didn't have one will cause the controller to [recompute the pool version](#pool-version), rendering all existing unclaimed clusters stale, causing them to be replaced gradually.
+
+#### Adding An Entry to the Inventory
+If `MaxSize` is unset or less than/equal to the length of the inventory, and `Size` allows, adding a new entry will cause a new ClusterDeployment to be added to the pool.
+If the pool is already at `[Max]Size` there is no immediate effect.
+
+#### Removing An Entry to the Inventory
+- If the entry is unused (no pool CD with that name exists), this is a no-op.
+- If an _unclaimed_ CD exists with that name, we delete it.
+  The controller will replace it, assuming an unused entry is available.
+- If a _claimed_ CD exists with that name, no-op.
+
+These are conceived to correlate as closely as possible to what happens when editing a pool's `Size`.
+
+#### Deleting The Inventory
+This will change the [pool version](#pool-version), rendering existing unclaimed clusters stale and causing the controller to replace them gradually.
+The administrator may wish to speed up this process by manually deleting CDs, or scaling the pool size to zero and back.
+
+### Maintaining the lease of the InstallConfigPatch
+If two controller pods trying to build a ClusterDeployment for the ClusterPool end up fetching the same installConfigPatch, they will be trying to claim same configured DNS or a hostname - a chaos scenario. To avoid it, we need to maintain a lease for each InstallConfigPatch
+
+To solve the above problem, we will have `status.clusterDeploymentRef` field and `PatchAvailable` condition on InstallConfigPatch.
+- When the `PatchAvailable` condition is false and `status.clusterDeploymentRef` is nil, the InstallConfigPatch hasn't been claimed yet.
+- When we need to build a new ClusterDeployment, we
+    - Load up the InstallConfigPatch list from the inventory.
+    - Do a GET on each entry in the list and check if the `PatchAvailable` condition is true
+    - For the first InstallConfigPatch that is available, we will use the json patch in the `spec.patch` field to update the default install config generated by the clusterpool.
+    - Update the `spec.clusterPoolReference.installConfigPatchRef`field in the ClusterDeployment with a reference to InstallConfigPatch
+    - Post an update to the InstallConfigPatch with the reference to the ClusterDeployment. Also update the `PatchAvailable` condition to false.
+      Here's where we bounce if two controllers try to grab the same InstallConfigPatch: the loser will bounce the update with a 409; we'll requeue and start over.
+    - Proceed with existing algorithm to create new cluster.
+- Add a top-level routine to ensure all unavailable InstallConfigPatches have valid ClusterDeployments associated with them.
+  If not, attempt to create them through the same flow.
```

Comment:
...within the bounds of the various pool capacities.

(This should be interesting to figure out. I think we would end up slotting this into [this switch statement](https://github.com/openshift/hive/blob/d66896015d990af5d33503f6f7539bc28a7a2c17/pkg/controller/clusterpool/clusterpool_controller.go#L375), possibly embedded in [this case](https://github.com/openshift/hive/blob/d66896015d990af5d33503f6f7539bc28a7a2c17/pkg/controller/clusterpool/clusterpool_controller.go#L383-L388).)

Reply:
should there be a finalizer on ClusterDeployment that prevents it from getting deleted before releasing the lease on InstallConfigPatch?